### DISCUSSION TITLE
Word level timestamps and hallucinations

### DISCUSSION THREAD

**Discussion by mu4farooqi** [openai/whisper#813] [Show and tell]
I tried to integrate the amazing work which @jongwook did with [multilingual ASR notebook](https://github.com/openai/whisper/blob/main/notebooks/Multilingual_ASR.ipynb) into `transcribe` function of my local [repository](https://github.com/mu4farooqi/whisper/tree/word_level_ts). You can clone and try it.

I always remove silences using [silero-vad](https://github.com/snakers4/silero-vad) before feeding the input to Whisper. This helps avoiding hallucinations in all other chunks except the last one. This word level timestamps algorithm also removes hallucinations in the last chunk. For example if a last chunk is 6s long, hallucinated words would have same `begin` and `end` timestamps.

**Comment by Bumicom** [comment]
When I try the ASR notebook I get an error at last step.

The dimensions of the returned tensor is only two where four are expected.
Did you forget to checkin a file?


torch.Size([1152, 1024])
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[37], line 21
     19     weights = torch.concatenate(QKs)  # layers * heads * tokens * frames    
     20     print(weights.shape)
---> 21     weights = weights[:, :, :, : duration // AUDIO_SAMPLES_PER_TOKEN].cpu()
     22 #     weights = medfilt(weights, (1, 1, 1, medfilt_width))
     23 #     weights = torch.tensor(weights * qk_scale).softmax(dim=-1)
     24     
   (...)
     86 #     display(pd.DataFrame(data))
     87 #     display(HTML("<hr>"))

IndexError: too many indices for tensor of dimension 2

**Reply by mu4farooqi** [reply]
I didn’t write the notebook. I used the concepts introduced in notebook and integrated in transcribe function of my forked Whisper repo. You can try that if you want.

**Comment by eschmidbauer** [comment]
Hi @mu4farooqi 
Tested out your branch and it's fantastic, nice work!
One problem i noticed is it sometimes skips the first segment of audio
I found that the line causing it is [here](https://github.com/mu4farooqi/whisper/commit/fd7bf16f4d43d31ee11288b79a9e8a4d34df8b0b#diff-f6accbbb4ebcd3dd6815bf012490d9ba37eb89a65f2124adc95c2a39bc6941b7R310)

I understand you are doing work to prevent hallucinations there but it maybe causing an edge case with the first segment. Any ideas?

**Reply by mu4farooqi** [reply]
Do you have a test audio/video file?

**Reply by eschmidbauer** [reply]
i have an audio example but I cannot share it unfortunately.
I will try using some public audio to reproduce the issue

**Reply by eschmidbauer** [reply]
i found a public wav that reproduces it. Here is a log from my code highlighting issue:
```{"message": "empty word timestamps in segment", "id": 3, "segment": {"id": 3, "seek": 9000, "start": 90.0, "end": 116.08, "text": " to hear that people are not listening to us. These things are so disrespectful I can't even begin to say, you know what, it's wrong on so many levels. I'm not overdramatic and I'm not overreacting. These things happen and we can't allow these kinds of behaviors within our company. You know what, let me tell you a little story. Only three years ago this company got like three", "tokens": [50364, 281, 1568, 300, 561, 366, 406, 4764, 281, 505, 13, 1981, 721, 366, 370, 47750, 286, 393, 380, 754, 50688, 50688, 1841, 281, 584, 11, 291, 458, 437, 11, 309, 311, 2085, 322, 370, 867, 4358, 13, 286, 478, 406, 670, 15125, 293, 286, 478, 406, 51020, 51020, 670, 265, 41090, 13, 1981, 721, 1051, 293, 321, 393, 380, 2089, 613, 3685, 295, 15501, 1951, 527, 2237, 13, 51360, 51360, 509, 458, 437, 11, 718, 385, 980, 291, 257, 707, 1657, 13, 5686, 1045, 924, 2057, 341, 2237, 658, 411, 1045, 51668, 51668], "temperature": 0.0, "avg_logprob": -0.42213769210012336, "compression_ratio": 1.668141592920354, "no_speech_prob": 9.311522444477305e-05, "word_timestamps": []}}```
[input.wav.zip](https://github.com/openai/whisper/files/10355217/input.wav.zip)

**Reply by mu4farooqi** [reply]
It works for me using large-v2 model. I think you are not pre-processing the file using VAD. Remove chunks of silence with VAD and then try. I got perfect results with your audio. I'm using [silero-vad](https://github.com/snakers4/silero-vad) to remove silences.

Edit: I just updated the code in my branch to pre-process the file using silero-vad. You can also use pyannote VAD pipeline if it works better for your usecase.

**Reply by ryanheise** [reply]
If you pre-process the file using VAD, the word timestamps will no longer correspond to the original audio. You would need to also do some post-processing to offset the word timestamps by the length of audio that was deleted before each timestamp using the retained `speech_timestamps`. However, beware that when you compact all the speech segments together with no gaps in between, if the first word of a segment has a timestamp that is slightly early (since the word timestamps are still approximate), such an approach could incorrectly infer that word to belong to the previous speech segment because all the neighboring segments are tightly packed together. You might be able to avoid this by adding a bit more padding around each segment in `speech_timestamps` before you slice up the audio. This might also improve the transcription results in cases where the VAD is clipping things a bit too tightly. I would also suggest lowering the speech threshold for VAD from 0.5 to say 0.3 to ensure that the VAD doesn't cut out audio where there was actually some low level speech.

Also some unrelated notes:

- the dtw-python library is GPL and cannot be used with whisper's current license.
- this code assumes cuda. It has no fallback for cpu.

**Reply by mu4farooqi** [reply]
Thanks for the feedback.

-- I agree about padding. VAD has a minimum threshold of 100ms to separate two segments. I can probably use something less than 100ms (50ms may be) to avoid the scenario which you just described.
-- For timestamps yeah I need to use some mapping between old and new timestamps.
-- VAD threshold is 0.5 by default. I'm using 0.6 just to match whisper's threshold. I'll tweak after testing on number of test-cases.

**Reply by ryanheise** [reply]
> -- I agree about padding. VAD has a minimum threshold of 100ms to separate two segments. I can probably use something less than 100ms (50ms may be) to avoid the scenario which you just described.

I'm talking about something different here. The goal is to increase the minimum distance between segment to reduce the chances of an inferred word timestamp incorrectly crossing the boundary into another closely cramped segment. The goal is to introduce a minimum gap X between the last word of the previous segment and the first word of the next segment in the audio that is fed into Whisper. And by minimum gap, we are saying that X should be "large enough" to space out these segments.. If you reduce the threshold that you refer to, that will not guarantee that segments will be at least X seconds apart. So I would leave that particular threshold alone, but then after `speech_timestamps` is generated, you could add an extra X onto the end timestamp of each entry in that list. Then in a second pass, you could merge any adjacent overlapping segments in `speech_timestamps`. Then pass that list into silero's utility function to slice up the audio based on those new timestamps, and now the audio should have a sufficient gap between each segment to reduce the likelyhood of inaccurate inferred word timestamps crossing the boundary into their neighboring segment.

> -- For timestamps yeah I need to use some mapping between old and new timestamps. -- VAD threshold is 0.5 by default. I'm using 0.6 just to match whisper's threshold. I'll tweak after testing on number of test-cases.

Note that this threshold doesn't necessarily map to Whisper's own threshold. But a lower number here is better to pick up on low level speech. A low number will also have the effect of producing some longer segments, e.g. previously you might have detected two segments close to each other, whereas with a lower threshold, the fricative sounds etc. at the beginnings and endings of words would be more likely to be detected extending the boundaries of each segment and making them bleed into neighboring segments if they are close enough. But that's fine because they are close enough anyway to prevent the problems you were trying to prevent in Whisper with hallucinations etc. Setting a lower threshold will still pick up the longer gaps which are the ones you really want to target.

**Reply by mu4farooqi** [reply]
I got your point. What I was trying to say was: VAD only separates two segments if there is a silence for more than 100ms. In other words, segments processed by VAD may have sub-segments with 100ms silence zones.

We can use something close to 100ms to combine the segments so that word timestamps are as accurate as possible.

For VAD threshold, I agree this is not same as Whisper VAD threshold and setting a lower value would help. Setting a very low value e.g. 0.2-0.3s may potentially introduce hallucinations in middle segments. 0.45-0.5 makes more sense IMO.

**Reply by ilarth** [reply]
Regarding a compatible dtw library. Perhaps librosa? I don't know much of anything but the tl;dr on github says:  

> ISC license is functionally equivalent to the BSD 2-Clause and MIT licenses  

The library itself supports pre-computed distance matrix as used by the notebook making it easy to swap out the dtw call.

**Reply by ryanheise** [reply]
> I got your point. What I was trying to say was: VAD only separates two segments if there is a silence for more than 100ms. In other words, segments processed by VAD may have sub-segments with 100ms silence zones.
> 
> We can use something close to 100ms to combine the segments so that word timestamps are as accurate as possible.

There is no need to cut every single micro-gap for this purpose, only the large gaps, e.g. the gaps longer than 1 second. If two segments are already very close together, there's no need to cut out the micro gap between them.

Also, the problem I described of being able to reconstitute the correct timestamps in post-processing only applies to where you actually did cut out a gap, so it actually helps if you aren't over zealous in cutting out too many gaps if those gaps were unnecessary to cut out in the first place.

Happily, it turns out that the most difficult gaps to reconstitute are the small ones *and* it is also the small ones that we don't really need to care about cutting in the first place, so you only really need to care about cutting the large gaps.

> For VAD threshold, I agree this is not same as Whisper VAD threshold and setting a lower value would help. Setting a very low value e.g. 0.2-0.3s may potentially introduce hallucinations in middle segments. 0.45-0.5 makes more sense IMO.

I think it would be fine to have a parameter for it. I deal with audio where the intro includes some words over music and VAD picks it up only at a 0.3.

**Reply by dgoryeo** [reply]
> I think it would be fine to have a parameter for it. I deal with audio where the intro includes some words over music and VAD picks it up only at a 0.3.


@ryanheise , are there any methods, tools, rules-of-thumb that can help to pick the right VAD settings for an audio? I am working on an audio that has in parts speech-over-music, and in parts quiet dialogue. I am trying to figure out what VAD settings would help. Side note: I am using Silero-VAD.

**Reply by ryanheise** [reply]
If you're just using silero-vad as is, then it is a matter of experimenting with the various parameters on your own audio. But if you're asking about methods, one thing you can do is to dynamically adjust the parameters over time. Let's say that in some parts of the audio, 0.3 is the best threshold while in other parts of the audio, 0.5 is the best threshold, and you could decide the appropriate threshold based on metrics you're tracking in the audio in that region such as the distribution of energy or VAD levels in that region.

**Reply by mu4farooqi** [reply]
Can't we do something like: Have two threshold e.g. 0.5 and 0.3. And if VAD outputs maybe X% more segments than when using 0.5 then we use the output of 0.3 otherwise stick with 0.5.

**Reply by ryanheise** [reply]
I don't like the idea of running the inference twice and using twice as many resources. One other alternative is to run it only once with the lower threshold (0.3) to ensure that no low-level speech is ever sliced out, but then use the raw speech probabilities from silero-vad to come up with a heuristic to guess which inferred words might be hallucinations. The raw speech probs would then give you a continuous speech probability estimate over time with values between 0.3 to 1.0. The problem is that I don't think having words inferred in a region with a speech probability below 0.5 necessarily means it's a hallucination, because I get low-level speech in intros with music.

So all I'd want is just a very basic parameter for this threshold, and if you did want to do something more dynamic, that could be another parameter to optionally enable it.

If I were to do something dynamic, I would use the approach I previously described since an audio file may have parts with speech over music, and other parts with just speech over silence, and you might find different thresholds work best on different parts of the file, rather than a single threshold working best on the whole file.

**Reply by dgoryeo** [reply]
Would an approach/solution like this be ideal:

- preprocess audio to detect scenes/clips;
- classify each scene for audio characteristics: music, background, volume/loudeness, etc.;
- determine VAD (and Whisper) settings for each audio scene/clip based on their classification;
- process each scene/clip with assigned settings.

Would such an approach be feasible or would it be too resource hungry with little to gain? Thoughts?

**Reply by skanda1005** [reply]
> If you pre-process the file using VAD, the word timestamps will no longer correspond to the original audio. You would need to also do some post-processing to offset the word timestamps by the length of audio that was deleted before each timestamp using the retained `speech_timestamps`. However, beware that when you compact all the speech segments together with no gaps in between, if the first word of a segment has a timestamp that is slightly early (since the word timestamps are still approximate), such an approach could incorrectly infer that word to belong to the previous speech segment because all the neighboring segments are tightly packed together. You might be able to avoid this by adding a bit more padding around each segment in `speech_timestamps` before you slice up the audio. This might also improve the transcription results in cases where the VAD is clipping things a bit too tightly. I would also suggest lowering the speech threshold for VAD from 0.5 to say 0.3 to ensure that the VAD doesn't cut out audio where there was actually some low level speech.
> 
> Also some unrelated notes:
> 
> * the dtw-python library is GPL and cannot be used with whisper's current license.
> * this code assumes cuda. It has no fallback for cpu.

Hi, by any chance do you know how to do the preprocessing to get back the original timestamps of my original audio, this would solve a major issue I am facing.
Thanks in advance

**Reply by mu4farooqi** [reply]
To reconstruct timestamps, I can give you some idea how we can do it. First while you are doing VAD. You need to construct a map from new timestamps to old timestamps. Something like following. In this code, tss will look like something [(s1, e1), (s2, e2), ...] which you'll get from VAD.

```
def get_ts_mapping(tss, wav):
 '''
 tss: List of tuples of start and end timestamps from the original audio.
 wav: Audio file loaded using any audio loading lib i.e. torch-audio
 '''
  # To reconstruct original word-level timestamps we need this mapping.
  tss_mapping, chunks_len = [], 0
  
  for i in tss:
    s, e = int(i['start'] * SAMPLE_RATE), int(i['end'] * SAMPLE_RATE)
    s, e = min(len(wav), s), min(len(wav), e)
    chunks_len += e - s
    tss_mapping.append((i['end'], round(chunks_len / SAMPLE_RATE, 1)))

  return tss_mapping
```

Once you have word level timestamps, you can use above mapping to reconstruct original word level timestamps using binary search.

```
def get_corrected_word_dict(word, start, end, vad_tss_map):
  """
  Binary search for the chunk with the smallest timestamp greater than or equal to ts
  """

  if start > end:
    return dict(text=word, start=round(end, 2), end=round(end, 2))

  def bs(value):
    s, e = 0, len(vad_tss_map) - 1
    while s < e:
      m = s + (e - s) // 2
      if value <= vad_tss_map[m][1]:
        e = m
      else:
        s = m + 1

    return s

  start_idx = bs(start)
  corrected_begin = start + (
    (vad_tss_map[start_idx][0] -
     vad_tss_map[start_idx][1]) if len(vad_tss_map) > 0 else 0)

  end_idx = bs(end)
  corrected_end = end + (
    (vad_tss_map[end_idx][0] -
     vad_tss_map[end_idx][1]) if len(vad_tss_map) > 0 else 0)
  return dict(text=word, start=round(corrected_begin, 2), end=round(corrected_end, 2))
```

If may be hard to understand but it works according to my tests. Let. me know if you need more information.

**Reply by skanda1005** [reply]
Great! Thank you so much! Will give it a try!

**Reply by skanda1005** [reply]
There seems to be a small bug in your code, in the get_ts_mapping function it needs to be len(wav[0]) not len(wav).

Also, I am having difficulties in running the second function, could you help me in how I could run the binary search function?

edit: I got it all working! Thanks for your help!

**Reply by mayeaux** [reply]
I have a Javascript implementation of adjusting the timestamps as well, I'll probably post it as a Gist once it's cleaned up.

**Comment by mu4farooqi** [comment]
It's the original audio.
    On Tuesday, 21 February 2023 at 11:25:33 am GMT-5, Skanda Subramanyan ***@***.***> wrote:  
 
 


Just a clarification, in the first function, will the audio input be the original audio or will it be the modified audio after VAD?

**Comment by rodrigoGA** [comment]
Excuse my ignorance, but looking at @jongwook  notebook and your repository, I am not clear on the necessary changes.
As I understand from the comment, the notebooks recommends removing the fragments with repeated timestamps that are found at the end of the audio. 
Is this alone sufficient to eliminate the hallucinations (once VAD is applied)? 
Is there no correlation between temperature, avg_logprob, compression_ratio, and no_speech_prob that should also be taken into account to remove  the hallucinations?

**Comment by LaurinmyReha** [comment]
maybe this is useful for someone here still:
I wrote a article about this
https://nyra-labs.com/research/killing-hallucinations
which improves this behaviour quite drastically. Feel free to try out the model too :)
https://huggingface.co/nyralabs/CrisperWhisper2.0_large
